src/Features/Core/Portable/ConvertTupleToStruct/AbstractConvertTupleToStructCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,865 @@
+                var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+
+                var generator = project.LanguageServices.GetService<SyntaxGenerator>();
+
+                // Get the fully qualified name for the new type we're creating.  We'll use this
+                // at replacement points so that we can find the right type even if we're in a 
+                // different namespace.
+
+                // If the struct is being injected into the global namespace, then reference it with
+                // "global::NewStruct",  Otherwise, get the full name to the namespace, and append
+                // the NewStruct name to it.
+                var structNameNode = CreateStructNameNode(
+                    generator, structName, typeParameters, addRenameAnnotation: false);
+
+                var fullTypeName = containingNamespace.IsGlobalNamespace
+                    ? (TNameSyntax)generator.GlobalAliasedName(structNameNode)
+                    : (TNameSyntax)generator.QualifiedName(generator.NameExpression(containingNamespace), structNameNode);
+
+                fullTypeName = fullTypeName.WithAdditionalAnnotations(Simplifier.Annotation)
+                                           .WithAdditionalAnnotations(DoNotAllowVarAnnotation.Annotation);
+
+                foreach (var documentToUpdate in group)
+                {
+                    var document = documentToUpdate.Document;
+                    var syntaxRoot = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                    var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+                    // If we were given specific nodes to update, only update those.  Otherwise
+                    // updated everything from the root down.
+                    var nodesToUpdate = documentToUpdate.NodesToUpdate.IsDefaultOrEmpty
+                        ? ImmutableArray.Create(syntaxRoot)
+                        : documentToUpdate.NodesToUpdate;
+
+                    var editor = new SyntaxEditor(syntaxRoot, generator);
+
+                    var replaced = false;
+
+                    foreach (var container in nodesToUpdate)
+                    {
+                        replaced |= await ReplaceTupleExpressionsAndTypesInDocumentAsync(
+                            document, editor, tupleExprOrTypeNode, tupleType,
+                            fullTypeName, structName, typeParameters,
+                            container, cancellationToken).ConfigureAwait(false);
+                    }
+
+                    if (replaced)
+                    {
+                        // We made a replacement.  Keep track of this so we can update our solution
+                        // later.
+                        documentToEditorMap.Add(document, editor);
+                    }
+                }
+
+                GC.KeepAlive(compilation);
```

Comment:
I didn't understand this. We're still holding on to the compilation, so it should not be GC'ed, right?

Reply:
I'll document.  but this is subtle, and something the IDE sometimes needs to do.  We're processing some subset of documents D in a project.  As we process each document we'll get its semanitc model.  But between documents, there is nothing stopping the system from collecting the compilation that semantic-model was using.  That means when we go to the next document, we may have to create its' semantic model *and* recreate the compilation for the project again.

To help avoid that, we 'hold onto' the compilation at the project level so it doesn't go away.

Reply:
Note: to be specific:

> We're still holding on to the compilation,

having a local point at something doesn't mean the GC will actually not collect it.  The GC is smart and it can see if you dont' have any references to the local then they will not root the object.